src/common/message-creators/action-message-dispatcher.ts

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { PayloadWithEventName } from '../../background/actions/action-payloads';
+import { Message } from '../message';
+import { Messages } from '../messages';
+import { TelemetryData } from '../telemetry-events';
+
+export class ActionMessageDispatcher {
+    constructor(private postMessageDelegate: (message: Message) => void, private tabId: number) {}
+
+    public dispatchMessage(message: Message): void {
+        if (this.tabId != null) {
+            message.tabId = this.tabId;
```

Comment:
Based on the naming, it was surprising to me that this function would mutate its input. Consider `{ ...message, this.tabId }` instead

Reply:
Done. I added a related type: `InterpreterMessage` that "extend" `Message` and add an optional `tabId` property. Now `Message` doesn't have the `tabId?` property. In a following PR I'll rename `Message` to `ActionMessage` to match the naming of our `XActionMessageCreator`'s